concordium-consensus/src/Concordium/KonsensusV1/Transactions.hs

```diff
@@ -186,50 +201,57 @@ processBlockItems ::
     BlockPointer pv ->
     -- |Return 'True' only if all transactions were
     -- successfully processed otherwise 'False'.
-    m Bool
-processBlockItems bb parentPointer = process $! Vector.toList $ bbTransactions bb
+    m (Maybe [(BlockItem, TVer.VerificationResult)])
+processBlockItems bb parentPointer = do
+    verificationContext <- getCtx
+    runMaybeT $ mapM (process verificationContext) $ Vector.toList $ bbTransactions bb
   where
     -- Create a context suitable for verifying a transaction within a 'Block' context.
     getCtx = do
-        _ctxSkovData <- get
         let _ctxBs = bpState parentPointer
+        -- We base the block energy limit on the chain parameters for the parent block.
+        -- This might not be accurate, because the parameter can change, but exceeding the limit
+        -- is treated as 'MaybeOK', so we will not fail even if the check is incorrect.
+        -- When the block is actually executed, the actual energy limit will be enforced.
         chainParams <- Concordium.GlobalState.BlockState.getChainParameters _ctxBs
         let _ctxMaxBlockEnergy = chainParams ^. cpConsensusParameters . cpBlockEnergyLimit
-        return $! Context{_ctxTransactionOrigin = TVer.Block, ..}
+        return Context{_ctxTransactionOrigin = TVer.Block, ..}
     theRound = bbRound bb
     theTime = bbTimestamp bb
-    -- Process the list of transactions recursively.
-    process :: [BlockItem] -> m Bool
-    process [] = return True
-    process (bi : bis) = do
+    -- Process a transaction
+    process ::
+        Context (PBS.HashedPersistentBlockState pv) ->
+        BlockItem ->
+        MaybeT m (BlockItem, TVer.VerificationResult)
```

Comment:
Generally MaybeT is not ideal in a tight loop, which this partially is. The bind is rather expensive due to constant checking of Just/Nothign.